text/0000-proc-macros.md

```diff
@@ -0,0 +1,417 @@
+specific signature and attribute. Where `foo` is the name of a function-like
+macro:
+
+```
+#[macro]
+pub fn foo(TokenStream, &mut MacroContext) -> TokenStream;
+```
+
+The first argument is the tokens between the delimiters in the macro use.
+For example in `foo!(a, b, c)`, the first argument would be `[Ident(a), Comma,
+Ident(b), Comma, Ident(c)]`.
+
+The value returned replaces the macro use.
+
+Attribute-like:
+
+```
+#[macro_attribute]
+pub fn foo(Option<TokenStream>, TokenStream, &mut MacroContext) -> TokenStream;
+```
+
+The first argument is a list of the tokens between the delimiters in the macro
+use. Examples:
+
+* `#[foo]` => `None`
+* `#[foo()]` => `Some([])`
+* `#[foo(a, b, c)]` => `Some([Ident(a), Comma, Ident(b), Comma, Ident(c)])`
+
+The second argument is the tokens for the AST node the attribute is placed on.
+Note that in order to compute the tokens to pass here, the compiler must be able
+to parse the code the attribute is applied to. However, the AST for the node
+passed to the macro is discarded, it is not passed to the macro nor used by the
+compiler (in practice, this might not be 100% true due to optimisiations). If
+the macro wants an AST, it must parse the tokens itself.
+
+The attribute and the AST node it is applied to are both replaced by the
+returned tokens. In most cases, the tokens returned by a procedural macro will
+be parsed by the compiler. It is the procedural macro's responsibility to ensure
+that the tokens parse without error. In some cases, the tokens will be consumed
+by another macro without parsing, in which case they do not need to parse. The
+distinction is not statically enforced. It could be, but I don't think the
+overhead would be justified.
+
+We also introduce a special configuration option: `#[cfg(macro)]`. Items with
+this configuration are not macros themselves but are compiled only for macro
+uses.
+
+Initially, it will only be legal to apply `#[cfg(macro)]` to a whole crate and
+the `#[macro]` and `#[macro_attribute]` attributes may only appear within a
+`#[cfg(macro)]` crate. This has the effect of partitioning crates into macro-
+defining and non-macro defining crates. Macros may not be used in the crate in
+which they are defined, although they may be called as regular functions. In the
+future, I hope we can relax these restrictions so that macro and non-macro code
+can live in the same crate.
+
+Importing macros for use means using `extern crate` to make the crate available
+and then using `use` imports or paths to name macros, just like other items.
+Again, see [RFC 1561](https://github.com/rust-lang/rfcs/pull/1561) for more
+details.
+
+When a `#[cfg(macro)]` crate is `extern crate`ed, it's items (even public ones)
+are not available to the importing crate; only macros declared in that crate.
+The crate is dynamically linked with the compiler at compile-time, rather
+than with the importing crate at runtime.
```

Comment:
This should come in hand with a warning/lint about adding public items in a `#[cfg(macro)]` crate